hbase-server/src/main/java/org/apache/hadoop/hbase/wal/AbstractWALProvider.java

```diff
@@ -150,11 +161,26 @@ private WAL getRemoteWAL(RegionInfo region, String peerId, String remoteWALDir)
       WAL wal = createRemoteWAL(region, ReplicationUtils.getRemoteWALFileSystem(conf, remoteWALDir),
         ReplicationUtils.getPeerRemoteWALDir(remoteWALDir, peerId), getRemoteWALPrefix(peerId),
         ReplicationUtils.SYNC_WAL_SUFFIX);
+      numRemoteWALUnderCreationLock.lock();
+      try {
+        numRemoteWALUnderCreation++;
+      } finally {
+        numRemoteWALUnderCreationLock.unlock();
+      }
       initWAL(wal);
       peerId2WAL.put(peerId, Optional.of(wal));
```

Comment:
These code modifications look good to me, but I am not familiar with `AbstractWALProvider`. I have a question: What happens if a 'remoteWAL' is being constructed, but `getWALs()` doesn't return it? (The 'remoteWAL' being constructed has not been put into `peerId2WAL` yet.)

Reply:
You can see the analysis in this PR.

https://github.com/apache/hbase/pull/5443#issuecomment-1751996650

It will cause problem for replication. We rely on whether a WAL file is beingWritten to determine whether we can move on to the next file. If we miss a WAL here but the WAL file has been put into the replication queue, we could get beingWritten == false but actually the file is beingWritten, which causes unexpected behavior...